sw/device/lib/crypto/api.h

```diff
@@ -0,0 +1,1378 @@
+  // AES CTR mode (counter mode).
+  kBlockCipherModeCtr = 4
+} block_cipher_mode_t;
+
+/**
+ * Enum to define AES operation to be performed.
+ */
+typedef enum aes_operation {
+  // AES operation mode encrypt.
+  kAesOperationEncrypt = 0,
+  // AES operation mode decrypt.
+  kAesOperationDecrypt = 1
+} aes_operation_t;
+
+/**
+ * Enum to define padding scheme for AES data.
+ */
+typedef enum aes_padding {
+  // Pads with value same as the number of padding bytes.
+  kAesPaddingPkcs7 = 0,
+  // Pads with 0x80 (10000000), followed by zero bytes.
+  kAesPaddingIso9797M2 = 1,
+  // Pads with 0x00 bytes.
+  kAesPaddingIso9797M1 = 2,
+  // Pads with random bytes, last byte is no. of padded bytes.
+  kAesPaddingRandom = 3,
+  // Pads with 0x00 bytes, last byte is no. of padded bytes.
+  kAesPaddingX923 = 4,
+  // Add no padding.
+  kAesPaddingNull = 5
+} aes_padding_t;
+
+/**
+ * Enum to define hashing mode.
+ */
+typedef enum hash_mode {
+  // SHA2-256 mode.
+  kHashModeSha256 = 0,
+  // SHA2-384 mode.
+  kHashModeSha384 = 1,
+  // SHA2-512 mode.
+  kHashModeSha512 = 2,
+  // SHA3-224 mode.
+  kHashModeSha3_224 = 3,
+  // SHA3-256 mode.
+  kHashModeSha3_256 = 4,
+  // SHA3-384 mode.
+  kHashModeSha3_384 = 5,
+  // SHA3-512 mode.
+  kHashModeSha3_512 = 6,
+  // SHA3-Shake128 mode.
+  kHashModeSha3Shake128 = 7,
+  // SHA3-Shake256 mode.
+  kHashModeSha3Shake256 = 8,
+  // SHA3-cShake128 mode.
+  kHashModeSha3Cshake128 = 9,
+  // SHA3-cShake256 mode.
+  kHashModeSha3Cshake256 = 10
+} hash_mode_t;
+
+/**
+ * Enum to define MAC mode.
+ */
+typedef enum mac_mode {
+  // HMAC-SHA2-256 mode.
+  kMacModeHmacSha256 = 0,
+  // KMAC-128 mode.
+  kMacModeKmac128 = 1,
+  // KMAC-256 mode.
+  kMacModeKmac256 = 2
+} mac_mode_t;
+
+/**
+ * Enum to define padding scheme for RSA data.
+ */
+typedef enum rsa_padding {
+  // Pads input data according to the PKCS#1-OAEP scheme.
+  kRsaPaddingOaep = 0,
+  // Pads input data according to the PKCS#1 (v1.5) scheme.
+  kRsaPaddingPkcs = 1,
+  // Pads input data according to the PKCS#1-PSS scheme.
+  kRsaPaddingPss = 2,
+  // NULL padding.
+  kRsaPaddingNull = 3
+} rsa_padding_t;
+
+/**
+ * Enum to define hash modes for RSA scheme
+ *
+ * Aligning with existing HASH modes.
+ */
+typedef enum rsa_hash {
+  // SHA2-256 hashing mode for RSA.
+  kRsaHashSha256 = 0,
+  // SHA2-384 hashing mode for RSA.
+  kRsaHashSha384 = 1,
+  // SHA2-512 hashing mode for RSA.
+  kRsaHashSha512 = 2,
+  // SHA3-384 hashing mode for RSA.
+  kRsaHashSha3_384 = 3
+} rsa_hash_t;
+
+/**
+ * Enum to define named elliptic curves.
+ */
+typedef enum ecc_named_curve {
+  // ECC named curve NIST P256.
+  kEccNamedCurveNistP256 = 0,
+  // ECC named curve NIST P384.
+  kEccNamedCurveNistP384 = 1,
+  // ECC named curve brainpool P256r1.
+  kEccNamedCurveBrainpoolP256R1 = 2
+} ecc_named_curve_t;
+
+/**
+ * Struct to handle ECDSA result.
+ */
+typedef struct ecc_signature {
+  // Length of ECC signature R parameter, in bytes.
+  size_t signature_r_len;
```

Comment:
OOC, why do R and S have lengths, while the point in ecc_public_key don't?

Also, `signature_r` seems a bit long... given r and s are pretty canonical names for the ECDSA scalars, we should use just that? (ditto for ecc_public_key; `x` and `y` are fine names).